lib/wasi/src/runtime/task_manager/mod.rs

```diff
@@ -233,27 +233,45 @@ impl dyn VirtualTaskManager {
         task: Box<WasmResumeTask>,
         ctx: WasiFunctionEnv,
         mut store: Store,
+        process_signals: bool,
+        main_thread: bool,
         trigger: Pin<Box<AsyncifyFuture>>,
     ) -> Result<(), WasiThreadError> {
         // This poller will process any signals when the main working function is idle
         struct AsyncifyPollerOwned {
             thread: WasiThread,
+            process_signals: bool,
             trigger: Pin<Box<AsyncifyFuture>>,
         }
         impl Future for AsyncifyPollerOwned {
             type Output = Result<Bytes, ExitCode>;
             fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
                 let work = self.trigger.as_mut();
-                Poll::Ready(if let Poll::Ready(res) = work.poll(cx) {
-                    Ok(res)
-                } else if let Some(forced_exit) = self.thread.try_join() {
+
+                if let Poll::Ready(res) = work.poll(cx) {
+                    return Poll::Ready(Ok(res));
+                }
+                if self.process_signals && self.thread.has_signals_or_subscribe(cx.waker()) {
+                    let signals = self.thread.signals().lock().unwrap();
+                    for sig in signals.0.iter() {
+                        if *sig == Signal::Sigint
+                            || *sig == Signal::Sigquit
+                            || *sig == Signal::Sigkill
+                            || *sig == Signal::Sigabrt
```

Comment:
Should we use `signals.0.drain(..)` instead of `iter()` so we consume all the signals and don't try to handle them again?

```suggestion
                    let mut signals = self.thread.signals().lock().unwrap();
                    for sig in signals.0.drain(..) {
                        if sig == Signal::Sigint
                            || sig == Signal::Sigquit
                            || sig == Signal::Sigkill
                            || sig == Signal::Sigabrt
```

Reply:
we can't drain it as the signals cant actually be processed here, so we want to leave them there and only terminate if specific signals are waiting to be processed